ops/lib/tcp_load_balancer/tcp_lb.py

```diff
@@ -0,0 +1,241 @@
+                health_monitor_data = app_data.get('health_monitor')
+                if health_monitor_data is None:
+                    continue
+                health_monitor = pickle.loads(health_monitor_data.encode('utf-8'))
+
+                members = []
+                for unit in relation.units:
+                    member_data = relation.data[unit].get('member')
+                    if member_data is None:
+                        continue
+                    members.append(pickle.loads(member_data.encode('utf-8')))
+                self._member_pools.append(MemberPool(listener, members, health_monitor))
+        return self._member_pools
+
+
+class MemberPool:
+
+    '''A class that aggregates the information about a pool of load-balancer members'''
+
+    def __init__(self, listener, members, health_monitor=None):
+        self.members = members
+        self.listener = listener
+        self.health_monitor = health_monitor
+
+
+class LBAvailable(EventBase):
+
+    '''Event emitted by TcpLoadBalancer.on.lb_available.
+
+    This event will be emitted when a new load-balancer instance appears on a relation for a member
+    to expose its data. If a load-balancer is highly-available, there will be multiple of those
+    events fired as instances of the load-balancer are observed by the member.
+    '''
+    pass
+
+
+class TcpLoadBalancerEvents(EventSetBase):
+
+    '''Events emitted by the TcpLoadBalancer class.'''
+    lb_available = EventSource(LBAvailable)
+
+
+class TcpLoadBalancer(Object):
+
+    on = TcpLoadBalancerEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm, relation_name, lb_algorithm=None):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._lb_algorithm = lb_algorithm
+
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event):
+        self.on.lb_available.emit()
+
+    @property
+    def _relation(self):
+        # TODO: there could be multiple independent reverse proxies in theory, address that later.
+        return self.model.get_relation(self._relation_name)
+
+    def expose_member(self, member, listener, health_monitor):
+        our_unit_data = self._relation.data[self.model.unit]
+        if member.address is None:
+            addr = self.model.get_binding(self._relation).network.ingress_address
+            member.set_address(addr)
+
+        our_unit_data['member'] = pickle.dumps(member, 0).decode('utf-8')
+        if self.model.unit.is_leader():
+            our_app_data = self._relation.data[self.model.app]
+            our_app_data['listener'] = pickle.dumps(listener, 0).decode('utf-8')
+            # A monitor for a pool of members.
+            our_app_data['health_monitor'] = pickle.dumps(health_monitor, 0).decode('utf-8')
+            our_app_data['lb_algorithm'] = self._lb_algorithm
+
+
+class Listener:
```

Comment:
Consider using [dataclasses](https://docs.python.org/3/library/dataclasses.html) here and in `Member`, `HealthMonitor`and `HTTPHealthMonitor`. It will make the accepted types clearer and reduces the possibility of typos etc within the boilerplate.